pkg/client/devops/fake/fakeJenkinsClient.go

```diff
@@ -0,0 +1,72 @@
+package fake
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/jenkins-zh/jenkins-client/pkg/core"
+	"github.com/jenkins-zh/jenkins-client/pkg/job"
+	"github.com/jenkins-zh/jenkins-client/pkg/mock/mhttp"
+	"kubesphere.io/devops/pkg/client/devops/jclient"
+)
+
+func NewFakeJenkinsClient(roundTripper *mhttp.MockRoundTripper) (j *jclient.JenkinsClient) {
+	core := core.JenkinsCore{
+		URL:          "http://localhost",
+		UserName:     "",
+		Token:        "",
+		RoundTripper: roundTripper,
+	}
+	j = &jclient.JenkinsClient{
+		Core: core,
+	}
+	return
+}
+
+// PrepareForGetProjectPipeline404 only for test
+func PrepareForGetProjectPipeline404(roundTripper *mhttp.MockRoundTripper, rootURL, user, password string, path, name string) {
+	request, _ := http.NewRequest("GET", fmt.Sprintf("%s/job/%s/job/%s/api/json", rootURL, name, path), nil)
+	response := &http.Response{
+		StatusCode: 404,
+		Proto:      "HTTP/1.1",
+		Request:    request,
+		Body: ioutil.NopCloser(bytes.NewBufferString(`
+		{"name":""}
+		`)),
+	}
+	roundTripper.EXPECT().
+		RoundTrip(core.NewRequestMatcher(request)).Return(response, nil)
+}
+
+// PrepareForCreateProjectPipeline only for test
+func PrepareForCreateProjectPipeline(roundTripper *mhttp.MockRoundTripper, rootURL, user, password string, jobPayload job.CreateJobPayload, path string) {
+	PrepareForGetProjectPipeline404(roundTripper, rootURL, user, password, jobPayload.Name, path)
```

Comment:
Why invoke this method `PrepareForGetProjectPipeline404` here?

Reply:
Because `CreateProjectPipeline` will check whether the job name has been used before creating a job. If not, it will create a job. Method `PrepareForGetProjectPipeline404` gives a 404 response so that `CreateProjectPipeline` will know the job name is available and create a job.

Reply:
I'd prefer to invoke `PrepareForGetProjectPipeline404` inside test function as needed.

Reply:
I am sorry. I don't quite understand here. What does `inside test function` mean? Did you mean inside `TestCreateProjectPipeline`?